LeaderLogProcessor.invokeOnCompletion handler violates kotlinx contract (slow, blocking), allowing parent allocator to close mid-handler

`LeaderLogProcessor`'s `init` block registers an `invokeOnCompletion` handler on the term's `SupervisorJob` that performs slow synchronous I/O — `extSource?.close()` (Postgres / Kafka external-source teardown), `replicaProducer.close()` (Kafka transactional producer flush+close), then `allocator.close()` on the leader's child allocator. The kotlinx contract for `CompletionHandler` requires handlers to be **"fast, non-blocking, and thread-safe"** and explicitly permits them to **"be invoked concurrently with the surrounding code."** LP's handler violates that contract.

The symptom is `IllegalStateException: Attempting operation on allocator when allocator is closed` thrown from inside the handler at `LeaderLogProcessor.kt:357`, wrapped as `CompletionHandlerException` and reported via the database scope's `CoroutineExceptionHandler` as `"ingestion stopping"`.

The exception is intermittent in its consequences. Most fires are absorbed by the supervisor's exception handler and just spam the log. Occasionally one wedges the test thread enough that the Kafka consumer-coordinator loses its heartbeat lease and the 15-minute `Integration Test` CI cap kills the job with no visible progress.

## Symptoms

Recent integration-test runs intermittently fail with this pattern:

- Active test progress, leader/follower transitions logging cleanly.
- A multi-minute silence with no test output.
- A `[kafka-coordinator-heartbeat-thread] WARN ConsumerCoordinator: consumer poll timeout has expired` line.
- The 15-minute `Integration Test` GitHub Action cap kills the job.

Underneath, the test process is generating exceptions like:

```
Caused by: kotlinx.coroutines.CompletionHandlerException: Exception in completion handler
  InvokeOnCompletion@... for SupervisorJobImpl{Cancelled}@...
Caused by: java.lang.IllegalStateException: Attempting operation on allocator when allocator is closed.
  Allocator(database/xtdb) ...
  at org.apache.arrow.memory.BaseAllocator.assertOpen
  at org.apache.arrow.memory.BaseAllocator.childClosed
  at org.apache.arrow.memory.BaseAllocator.close
  at xtdb.indexer.LeaderLogProcessor._init_$lambda$1(LeaderLogProcessor.kt:357)
```

Recent integration-test runs on `main` consistently show 11–25 occurrences per run on both passing and failing builds — see [run 80364648047](https://github.com/xtdb/xtdb/actions/runs/27217943936/job/80364648047) (passed, 16 fires) and [80329198507](https://github.com/xtdb/xtdb/actions/runs/27208070468/job/80329198507) (failed, 14 fires) for a contrast pair. `FollowerLogProcessor.kt:293` has an equivalent `invokeOnCompletion { allocator.close() }` (though structurally different — registered on the tail-job rather than the supervisor); the counts above combine both sites.

## Evidence

### Bisect

Bisect across the `job-owned-service-lifecycle` merge confirms this is a regression introduced by that refactor:

| Commit | Integration test outcome | Exception occurrences in log |
|---|---|---|
| `37cc3efa4` (parent of merge) | success | 0 |
| `65fad51ce` (lifecycle merge) | success | many |

Pre-merge, the close path was sequential — `runBlocking { processor.cancelAndJoin() }` before `closeAll(allocator)`, with leader cleanup orchestrated by an explicit `suspend cancelAndJoin()` method. The `invokeOnCompletion`-based teardown was introduced by `5d5994c88 refactor(indexer): leader/follower processors own their lifetime`.

### Instrumented integration-test run

Instrumenting LP's handler with thread / coroutine / timestamp logs (worktree `worktree-5703-instrument`) and running the kafka-connect-source integration test produced the bug 18 times in a single run. The first occurrence captures the precise sequence:

```
16:03:12.268 [Test worker]   database[xtdb] close() ENTRY
16:03:12.269 [Test worker]   database[xtdb] before runBlocking { job.cancelAndJoin() }
16:03:12.269 [worker-7 #72]  leader[xtdb] handler ENTRY                                  ← term supervisor finalised, handler starts
16:03:12.269 [worker-7 #72]  leader[xtdb] step=closing extSource
16:03:12.269 [worker-7 #72]  leader[xtdb] step=closing replicaProducer
16:03:12.270 [Test worker]   database[xtdb] after runBlocking — before closeAll          ← runBlocking RETURNED
16:03:12.270 [Test worker]   database[xtdb] CLOSING database/xtdb allocator
16:03:12.270 [worker-7 #72]  leader[xtdb] step=closing leader-child allocator
16:03:12.271 [Test worker]   database[xtdb] database/xtdb allocator close OK             ← parent now closed
16:03:12.272 [worker-7 #72]  leader[xtdb] handler THREW IllegalStateException
```

The leader handler is mid-flight (extSource and producer closed, allocator not yet attempted) when `runBlocking { cancelAndJoin() }` returns and `closeAll` closes the database allocator. The handler then reaches its `allocator.close()` step and trips `BaseAllocator.assertOpen()`.

## Root cause — `tryWaitForChild` skips already-final children

The race is not "in principle possible per the kdoc" — it has a precise structural cause in `kotlinx-coroutines-core` `1.10.2`. Two facts compose:

**1. State CAS happens before handler iteration.** When a job finalises, its `_state` is CAS'd to the final state *before* `completeStateFinalization` walks the handler list. See [`finalizeFinishingState`](https://github.com/Kotlin/kotlinx.coroutines/blob/1.10.2/kotlinx-coroutines-core/common/src/JobSupport.kt) in JobSupport — `_state.compareAndSet(state, finalState.boxIncomplete())` precedes `completeStateFinalization(state, finalState)`. Any observer of state from this point on sees the job as "done" while its handlers are still iterating.

**2. A parent waiting on a child short-circuits on already-final state.** A parent learns about child completion only through `ChildCompletion`, registered by [`tryWaitForChild`](https://github.com/Kotlin/kotlinx.coroutines/blob/1.10.2/kotlinx-coroutines-core/common/src/JobSupport.kt):

```kotlin
private tailrec fun tryWaitForChild(state: Finishing, child: ChildHandleNode, proposedUpdate: Any?): Boolean {
    val handle = child.childJob.invokeOnCompletion(
        invokeImmediately = false,
        handler = ChildCompletion(this, state, child, proposedUpdate)
    )
    if (handle !== NonDisposableHandle) return true       // wait registered
    val nextChild = child.nextChild() ?: return false
    return tryWaitForChild(state, nextChild, proposedUpdate)  // child was already final — skip
}
```

`invokeOnCompletion(invokeImmediately = false)` returns `NonDisposableHandle` when the target's state is already final. `tryWaitForChild` interprets that as "this child is done" and moves on without waiting. **The parent never waits for the child's handlers to finish iterating.**

Compose the two: any time a child's state CASes to final before the parent's `tryWaitForChild` reaches it, the parent skips that child. Any `invokeOnCompletion` handler on the skipped child runs concurrently with whatever the parent does after `cancelAndJoin` returns.

The instrumented trace above shows this end-to-end:
1. `Database.close()` calls `runBlocking { job.cancelAndJoin() }`.
2. Cancellation cascades. The term supervisor's children (persister, extJob) all complete in microseconds — their `finally` blocks only `Channel.close()` synchronously.
3. The term supervisor's state CASes to final on the cascading thread. LP's `invokeOnCompletion` handler is dispatched onto worker-7 and begins its slow I/O (extSource close, producer close).
4. The cancelling thread reaches `tryWaitForChild` for the term supervisor, sees the supervisor's state as final, gets `NonDisposableHandle`, and skips. No `ChildCompletion` is registered.
5. With no incomplete children left, the database job finalises. `cancelAndJoin` returns on the test thread.
6. `closeAll` runs and closes `Allocator(database/xtdb)`.
7. On worker-7, LP's handler advances to `allocator.close()` on its leader-child allocator. `BaseAllocator.childClosed` walks up to the now-closed parent and throws.

`dev/doc/coroutines.adoc` (in this branch) documents the lifecycle semantics with the same source pointers for future reference.

## Proposed approach

The kotlinx contract is clear: `invokeOnCompletion` is for *notification* — fast, non-blocking, fire-and-forget. It is not a cleanup mechanism for resources whose teardown has timing requirements relative to surrounding code. From the [official Kotlin Cancellation and Timeouts docs](https://kotlinlang.org/docs/cancellation-and-timeouts.html), the endorsed idiom for resource teardown across cancellation is `try`/`finally` inside the coroutine body, with `withContext(NonCancellable)` wrapping any cleanup that itself needs to suspend:

```kotlin
suspend fun runLeaderTerm() = coroutineScope {
    val allocator = parent.newChildAllocator("leader-log-processor", ...)
    try {
        supervisorScope {                   // siblings don't cancel each other on failure
            launch { runPersister() }
            launch { runExtJob(extSource) }
            launch { runGc() }
        }
    } finally {
        withContext(NonCancellable) {       // cleanup completes even if we're being cancelled
            extSource?.close()
            replicaProducer.close()
            allocator.close()
        }
    }
}
```

The `try`/`finally` *is* the cleanup mechanism. The coroutine doesn't complete until `finally` returns, so any join — `cancelAndJoin` on the term, cascading cancellation from `Database.close()`, structural-concurrency child notification — properly waits. Cleanup is colocated with resource acquisition; there is no race window between "join returned" and "handler finished" because there is no handler.

`withContext(NonCancellable)` is documented at the URL above: *"NonCancellable is useful when you need to ensure that certain operations, such as closing resources with a suspending `close()` function, complete even if the coroutine is canceled."* For LP the closes are non-suspend (Postgres connection, Kafka producer, Arrow allocator), so `NonCancellable` is mostly precautionary — it ensures the cleanup completes even if any close call internally suspends or the coroutine is mid-cancel when `finally` runs.

Two kotlinx GitHub issues asked for a higher-level `Resource`-style wrapper API in the core library ([#1191](https://github.com/Kotlin/kotlinx.coroutines/issues/1191), [#1044](https://github.com/Kotlin/kotlinx.coroutines/issues/1044)). Both were closed without a wrapper being added — the maintainers' position is that try/finally with structured concurrency already gives correct semantics by construction. Ecosystem libraries like Arrow Kt's `resourceScope` package the same pattern; we don't use those.

### What this implies for the code

The current LP shape — `termSupervisor` with multiple independent child coroutines (`persisterJob`, `extJob`, `gcScope`) and an `invokeOnCompletion` on the supervisor — doesn't have a natural place to put a `finally`. The structural change required is: **make the term a single hosting coroutine** that uses `supervisorScope` internally to launch its siblings, with `try`/`finally` around the whole thing. The wrapper's `openTerm` becomes "launch the term coroutine" rather than "create SupervisorJob + construct LeaderLogProcessor".

`FollowerLogProcessor.kt:293`'s `invokeOnCompletion { allocator.close() }` is the same anti-pattern in miniature and should migrate the same way.

### Principle

`invokeOnCompletion` is for *notification*, not *cleanup*. Anything beyond a constant-time non-blocking action (updating a metric, recording a state change, disposing a `DisposableHandle`) belongs in the coroutine body's `try`/`finally`, not in a completion handler. The lifecycle refactor's other `invokeOnCompletion` uses are worth auditing against this principle while we're here.

## Progress

**Status**: completed

- [x] Identify the production code path that lets the race surface at the `Database.close()` layer — `tryWaitForChild` short-circuit on `NonDisposableHandle`, confirmed by instrumented integration-test trace and against kotlinx 1.10.2 source
- [x] Add dev doc on coroutine lifecycle (`dev/doc/coroutines.adoc`) — landed in #5706
- [x] Restructure `LeaderLogProcessor` (and `FollowerLogProcessor`) to use try/finally + `withContext(NonCancellable)` for cleanup — landed in #5706
- [x] Audit other `invokeOnCompletion` uses in the lifecycle refactor for the same anti-pattern — `Compactor.kt:206` is the only other instance; tracked separately as #5707
- [x] Confirm the exception count drops to zero in integration-test logs — verified locally (0 vs 18 pre-fix) and on CI ([run 27290148519](https://github.com/mbutlerw/xtdb/actions/runs/27290148519): 0 occurrences across all 9 jobs)

--- a/core/src/main/kotlin/xtdb/indexer/FollowerLogProcessor.kt
+++ b/core/src/main/kotlin/xtdb/indexer/FollowerLogProcessor.kt
@@ -5,7 +5,6 @@ import io.micrometer.core.instrument.MeterRegistry
 import io.micrometer.core.instrument.Timer
 import kotlinx.coroutines.CancellationException
 import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.launch
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.first
 import org.apache.arrow.memory.BufferAllocator
@@ -26,6 +25,7 @@ import xtdb.table.TableRef
 import xtdb.util.StringUtil.asLexHex
 import xtdb.util.debug
 import xtdb.util.error
+import xtdb.util.launchWithCleanup
 import xtdb.util.logger
 import xtdb.util.trace
 
@@ -276,12 +276,12 @@ class FollowerLogProcessor @JvmOverloads constructor(
         replicaState.value = ReplicaState.Failed(latestReplicaMsgId, e)
     }
 
-    // Launched last, so every field the tail touches is initialised before the loop's first record;
-    // `tailAll` suspends immediately on subscribe. A cancel (role transition / shutdown) stops the
-    // tail and, once it has joined, the completion handler frees this processor's child allocator.
-    // `allocator` is `this@…`-qualified: bare `allocator` in an init lambda binds the ctor param.
+    // Launched last so every field the tail touches is initialised before the first record.
+    // See dev/doc/coroutines.adoc / #5703 for why this uses launchWithCleanup rather than
+    // invokeOnCompletion. `allocator` is `this@…`-qualified: bare `allocator` in a launch lambda
+    // binds the ctor param.
     init {
-        val tailJob = scope.launch {
+        scope.launchWithCleanup(cleanup = { this@FollowerLogProcessor.allocator.close() }) {
             try {
                 replicaLog.tailAll(afterReplicaMsgId, this@FollowerLogProcessor)
             } catch (e: CancellationException) {
@@ -290,6 +290,5 @@ class FollowerLogProcessor @JvmOverloads constructor(
                 notifyError(e); throw e
             }
         }
-        tailJob.invokeOnCompletion { this@FollowerLogProcessor.allocator.close() }
     }
 }

--- a/core/src/main/kotlin/xtdb/indexer/LeaderLogProcessor.kt
+++ b/core/src/main/kotlin/xtdb/indexer/LeaderLogProcessor.kt
@@ -250,48 +250,6 @@ class LeaderLogProcessor(
         trieCatalog.deleteTries(task.tableName, task.trieKeys)
     }
 
-    // Unbiased select across the three sources so a slow producer on one channel can't starve
-    // the others. The three sources are: source-log records, external-source resolved txs, and
-    // GC trie deletions.
-    private val persisterJob: Job = scope.launch {
-        try {
-            while (true) {
-                val task: PersisterTask = selectUnbiased {
-                    sourceLogCh.onReceive { it }
-                    extSourceCh.onReceive { it }
-                    gcCh.onReceive { it }
-                }
-                try {
-                    when (task) {
-                        is SourceLogTask.Record -> handleSourceLogRecord(task)
-                        is SourceLogTask.ResolvedTx -> handleResolvedTx(task.resolvedTx, task.srcMsgId)
-                        is ExtSourceTask.ResolvedTx -> handleResolvedTx(task.resolvedTx, srcMsgId = null)
-                        is GcTask.TriesDeleted -> handleTriesDeleted(task)
-                    }
-                    task.onComplete.complete(Unit)
-                } catch (e: CancellationException) {
-                    task.onComplete.cancel(e)
-                    throw e
-                } catch (e: InterruptedException) {
-                    task.onComplete.completeExceptionally(e)
-                    throw e
-                } catch (e: Interrupted) {
-                    task.onComplete.completeExceptionally(e)
-                    throw e
-                } catch (e: Throwable) {
-                    watchers.notifyError(e)
-                    task.onComplete.completeExceptionally(e)
-                    throw e
-                }
-            }
-        } catch (_: Throwable) {
-        } finally {
-            sourceLogCh.close()
-            extSourceCh.close()
-            gcCh.close()
-        }
-    }
-
     private suspend fun submit(task: PersisterTask) {
         when (task) {
             is SourceLogTask -> sourceLogCh.send(task)
@@ -334,27 +292,70 @@ class LeaderLogProcessor(
 
     private val txErrorCounter: Counter? = nodeBase.meterRegistry?.let { Counter.builder("tx.error").register(it) }
 
-    private val extJob = extSource?.let { source ->
-        scope.launch {
-            try {
-                source.onPartitionAssigned(partition, watchers.externalSourceToken, this@LeaderLogProcessor)
-            } catch (e: CancellationException) {
-                throw e
-            } catch (e: Throwable) {
-                watchers.notifyError(e)
-            }
-        }
-    }
-
-    init {
-        // Cleanup rides the job tree, not an explicit teardown call: once the leader term's scope is
-        // cancelled and its coroutines (persister, ext source, GCs) have joined, free what this term
-        // opened — leaf-first. `allocator` must be `this@…`-qualified: bare `allocator` in an init
-        // lambda binds the ctor param (the parent), not this child field.
-        scope.coroutineContext.job.invokeOnCompletion {
+    // Hosts persister + the optional ext-source loop + cleanup. See dev/doc/coroutines.adoc /
+    // #5703 for why this uses launchWithCleanup rather than invokeOnCompletion.
+    private val termJob: Job = scope.launchWithCleanup(
+        cleanup = {
             extSource?.close()
             replicaProducer.close()
             this@LeaderLogProcessor.allocator.close()
+        }
+    ) {
+        // supervisorScope so an ext-source crash doesn't kill the persister.
+        supervisorScope {
+            // Persister: unbiased select across source-log records, external-source resolved
+            // txs, and GC trie deletions — so a slow producer on one channel can't starve the
+            // others.
+            launch {
+                try {
+                    while (true) {
+                        val task: PersisterTask = selectUnbiased {
+                            sourceLogCh.onReceive { it }
+                            extSourceCh.onReceive { it }
+                            gcCh.onReceive { it }
+                        }
+                        try {
+                            when (task) {
+                                is SourceLogTask.Record -> handleSourceLogRecord(task)
+                                is SourceLogTask.ResolvedTx -> handleResolvedTx(task.resolvedTx, task.srcMsgId)
+                                is ExtSourceTask.ResolvedTx -> handleResolvedTx(task.resolvedTx, srcMsgId = null)
+                                is GcTask.TriesDeleted -> handleTriesDeleted(task)
+                            }
+                            task.onComplete.complete(Unit)
+                        } catch (e: CancellationException) {
+                            task.onComplete.cancel(e)
+                            throw e
+                        } catch (e: InterruptedException) {
+                            task.onComplete.completeExceptionally(e)
+                            throw e
+                        } catch (e: Interrupted) {
+                            task.onComplete.completeExceptionally(e)
+                            throw e
+                        } catch (e: Throwable) {
+                            watchers.notifyError(e)
+                            task.onComplete.completeExceptionally(e)
+                            throw e
+                        }
+                    }
+                } catch (_: Throwable) {
+                } finally {
+                    sourceLogCh.close()
+                    extSourceCh.close()
+                    gcCh.close()
+                }
+            }
+
+            extSource?.let { source ->
+                launch {
+                    try {
+                        source.onPartitionAssigned(partition, watchers.externalSourceToken, this@LeaderLogProcessor)
+                    } catch (e: CancellationException) {
+                        throw e
+                    } catch (e: Throwable) {
+                        watchers.notifyError(e)
+                    }
+                }
+            }
         }
     }
 

--- a/core/src/main/kotlin/xtdb/util/CoroutineUtil.kt
+++ b/core/src/main/kotlin/xtdb/util/CoroutineUtil.kt
@@ -1,0 +1,37 @@
+package xtdb.util
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.NonCancellable
+import kotlinx.coroutines.awaitCancellation
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+/**
+ * Launches a child coroutine that runs [body], then runs [cleanup] under [NonCancellable].
+ *
+ * The cleanup is part of the launched Job's lifecycle: the Job doesn't reach a final state
+ * until cleanup has returned, so a parent's `cancelAndJoin` waits for it. That's the
+ * property [kotlinx.coroutines.Job.invokeOnCompletion] doesn't give you — see
+ * `dev/doc/coroutines.adoc` / xtdb#5703.
+ *
+ * Cleanup runs after the body whether the body returned normally, threw, or was cancelled.
+ * [NonCancellable] keeps cleanup running through cancellation requests that arrive while it
+ * is in flight.
+ *
+ * If [cleanup] throws, the Job ends `CompletedExceptionally` and the exception surfaces via
+ * the scope's [kotlinx.coroutines.CoroutineExceptionHandler].
+ *
+ * [body] defaults to [awaitCancellation] for the "exists only to run cleanup on cancel"
+ * shape; pass a body for "do work, then clean up after."
+ */
+fun CoroutineScope.launchWithCleanup(
+    cleanup: suspend () -> Unit,
+    body: suspend CoroutineScope.() -> Unit = { awaitCancellation() }
+): Job = launch {
+    try {
+        body()
+    } finally {
+        withContext(NonCancellable) { cleanup() }
+    }
+}

--- a/dev/doc/coroutines.adoc
+++ b/dev/doc/coroutines.adoc
@@ -1,0 +1,70 @@
+= Coroutines in XTDB
+
+:written: 2026-06-11
+
+Notes on coroutine patterns and gotchas specific to this codebase, supplementing the link:https://kotlinlang.org/docs/coroutines-overview.html[official Kotlin docs].
+We are on kotlinx-coroutines `1.10.2`; source links point there.
+
+== Patterns
+
+=== `launchWithCleanup` for cleanup that gates parent close
+
+For "do work, then close some things, and make sure a parent's `cancelAndJoin` waits for the close," use `xtdb.util.launchWithCleanup`.
+
+[source,kotlin]
+----
+scope.launchWithCleanup(
+    cleanup = {
+        extSource?.close()
+        replicaProducer.close()
+        allocator.close()
+    }
+) {
+    supervisorScope {
+        launch { runPersister() }
+        launch { runExt() }
+    }
+}
+----
+
+The helper's kdoc states the lifecycle property it gives you.
+In-tree users: `LeaderLogProcessor`, `FollowerLogProcessor`.
+
+Most reaches for `Job.invokeOnCompletion` to close resources should be this helper instead — see the gotcha below.
+
+== Gotchas
+
+=== `invokeOnCompletion` is not a fence over `cancelAndJoin`
+
+A `Job`'s state CASes to its final state *before* its `invokeOnCompletion` handlers iterate.
+Observers of the job's state — `join`, `isCompleted`, `invokeOnCompletion(invokeImmediately=false)` — see the job as "done" the moment the CAS lands, while handlers are still running on whatever thread triggered finalisation.
+
+In particular, a parent waiting on a child via `tryWaitForChild` short-circuits on a child whose state is already final:
+
+[source,kotlin]
+----
+private tailrec fun tryWaitForChild(state: Finishing, child: ChildHandleNode, ...): Boolean {
+    val handle = child.childJob.invokeOnCompletion(
+        invokeImmediately = false,
+        handler = ChildCompletion(this, state, child, ...)
+    )
+    if (handle !== NonDisposableHandle) return true   // <1>
+    val nextChild = child.nextChild() ?: return false  // <2>
+    return tryWaitForChild(state, nextChild, ...)
+}
+----
+<1> Child still incomplete — register the wait handler, return.
+<2> Child already final — the parent skips it without registering a wait.
+
+So if your `invokeOnCompletion` handler does any non-trivial work — slow I/O, multiple `.close()` calls — that work can race the parent's `cancelAndJoin` returning and whatever happens next.
+For cleanup that needs to gate a parent's close, put it in the coroutine body's `try/finally` instead (or use `launchWithCleanup`).
+
+Background: this was the root cause of link:https://github.com/xtdb/xtdb/issues/5703[#5703].
+
+== Source pointers
+
+* link:https://github.com/Kotlin/kotlinx.coroutines/blob/1.10.2/kotlinx-coroutines-core/common/src/JobSupport.kt[`JobSupport.kt`] — `Job` state machine.
+  Functions worth knowing: `finalizeFinishingState`, `completeStateFinalization`, `tryWaitForChild`, `continueCompleting`, `notifyCompletion`, `cancelImpl`.
+* link:https://github.com/Kotlin/kotlinx.coroutines/blob/1.10.2/kotlinx-coroutines-core/common/src/CompletionHandler.common.kt[`CompletionHandler.common.kt`] — `JobNode` hierarchy (`ChildHandleNode`, `ChildCompletion`, `ResumeOnCompletion`).
+* link:https://kotlinlang.org/docs/cancellation-and-timeouts.html[Kotlin: Cancellation and timeouts] — official guide.
+* link:https://elizarov.medium.com/structured-concurrency-722d765aa952[Structured concurrency] (Elizarov) — design intent behind parent/child Job relationships.

